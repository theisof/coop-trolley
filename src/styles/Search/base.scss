@import './search';
@import './search-item';