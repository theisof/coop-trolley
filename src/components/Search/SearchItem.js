import React, { Fragment } from 'react'

class SearchItem extends React.Component {
  constructor(props) {
    super(props);

    this.state = {}
    this.handleClick = this.handleClick.bind(this)
    this.handleKeyPress = this.handleKeyPress.bind(this)
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    return {
      ...prevState,
      id: `search-item__index-${nextProps.index}`
    }
  }

  componentDidMount(){
    document.addEventListener("keydown", this.handleKeyPress, false)
  }

  componentWillUnmount(){
    document.removeEventListener("keydown", this.handleKeyPress, false)
  }

  componentDidUpdate(prevProps) {
    if (this.props.isSelected && prevProps.isSelected !== this.props.isSelected) {
      document.getElementById(this.state.id).focus()
    }
  }

  handleKeyPress(e) {
    if (e.key === 'Enter' && this.props.isSelected) {
      e.preventDefault()
      this.handleClick()
    }
  }

  handleClick() {
    const { onSearchItemClick, link } = this.props

    if (link) {
      window.location.href = link
    } else {
      onSearchItemClick(this.props)
    }
  }

  render() {
    const { thumbnail, name, link, isSelected } = this.props
    const { id } = this.state
    const wrapperClass = isSelected ? 'coop-search__item coop-search__item--is-selected' : 'coop-search__item'
    const content = (
      <Fragment>
        {thumbnail &&
          <img src={thumbnail} alt={name} className='coop-search__thumbnail' />
        }

        <div  className='coop-search__name'>
          { name }
        </div>
      </Fragment>
    )

    if (link) {
      return (
        <a href={link} className={wrapperClass} id={id}>
          { content }
        </a>
      )
    }

    return (
      <div onClick={this.handleClick} className={wrapperClass} id={id}>
        { content }
      </div>
    )
  }
}

export default SearchItem
