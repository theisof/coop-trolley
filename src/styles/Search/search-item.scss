.coop-search__item {
  display: flex;
  align-items: center;
  padding: 5px;
  border-bottom: 1px solid #eee;
  cursor: pointer;

  &,
  &:hover,
  &:active,
  &:focus,
  &:visited {
    color: #000;
    text-decoration: none;
  }

  &:hover,
  &:active {
    .coop-search__thumbnail,
    .coop-search__name {
      opacity: 0.7;
    }
  }
}

.coop-search__item--is-selected {
  background-color: #555;

  &,
  &:hover,
  &:active,
  &:focus,
  &:visited {
    color: #fff;
  }
}

.coop-search__fallback {
  font-size: 15px;
  padding: 35px 15px;
  text-align: center;
  color: #555;
}

.coop-search__thumbnail {
  height: 35px;
  width: auto;
  max-width: 80px;
  margin: 2px 10px 2px 5px;
}

.coop-search__name {
  padding: 3px;
  font-size: 12px;
  line-height: 1.4;
}
