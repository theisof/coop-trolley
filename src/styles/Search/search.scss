.coop-search {
  position: relative;
  font-family: $font-standard;
  z-index: $z-index-backdrop + 1;
}

.coop-search__input {
  width: 100%;
  height: 35px;
  padding: 0 15px 0 38px;
  font-family: $font-standard;
  border: 1px solid $color-gray-light;
  font-size: 12px;

  &:focus {
    outline: 0;
  }
}

.coop-search__icon {
  position: absolute;
  height: 18px;
  top: 9px;
  left: 12px;
  opacity: 0.5;
  pointer-events: none;
}

.coop-search__results {
  position: absolute;
  width: 100%;
  background-color: #fff;
  z-index: $z-index-backdrop + 2;
  box-shadow: 0px 3px 5px 0px rgba(0, 0, 0, 0.25);
  max-height: calc(100vh - 80px);
  overflow: auto;
  -webkit-overflow-scrolling: touch;
}

.coop-search__section-title {
  font-size: 12px;
  font-weight: bold;
  padding: 10px;
  margin: 0;
  background-color: #eee;
  font-family: $font-display;
}

.coop-search__clear-wrap {
  width: 40px;
  right: 0;
  height: 100%;
  position: absolute;
  right: 0;
  z-index: $z-index-backdrop + 3;
  top: 0;
  align-items: center;
  justify-content: center;
  border-left: 1px solid #ddd;
  display: none;
  cursor: pointer;
}

.coop-search__clear-wrap--has-input {
  display: flex;
}

.coop-search__clear-image {
  width: 10px;
  height: 10px;
}

.coop-search__loader {
  right: 65px;
  left: auto;
}

@media (max-width: $breakpoint-sm) {
  .coop-search {
    height: $site-nav-height;
    border: none;
    padding: 0;
  }

  .coop-search__input {
    font-size: 16px;
    height: calc(100% + 1px);
    border-bottom: 1px solid #ddd;
    padding-left: 52px;
  }

  .coop-search__icon {
    height: 25px;
    top: 19px;
    left: 15px;
  }

  .coop-search__results {
    max-height: calc(100vh - #{$site-nav-height});
  }

  .coop-search__clear-wrap {
    width: 60px;
    display: flex;
  }

  .coop-search__clear-image {
    width: 15px;
    height: 15px;
  }

  .coop-search__loader {
    right: 85px;
  }
}