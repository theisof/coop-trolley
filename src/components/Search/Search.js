import React, { Fragment } from 'react'
import Backdrop from '../Backdrop'
import getImagePath from '../../utils/getImagePath'
import SearchItem from './SearchItem'
import Loader from 'react-loader'

class Search extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      searchText: '',
      selected: 0,
      didSearch: false,
      isActive: false
    }

    this.handleInputChange = this.handleInputChange.bind(this)
    this.hideSearch = this.hideSearch.bind(this)
    this.handleKeyPress = this.handleKeyPress.bind(this)
    this.handleBlur = this.handleBlur.bind(this)
    this.searchInput = React.createRef()
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    let count = 0
    let hasRequired = false

    const searchResults = nextProps.searchResults ? nextProps.searchResults.map( section => {
      if (! section.optional) {
        hasRequired = true
      }

      return {
        ...section,
        results: section.results.map( item => {
          count++

          return {
            ...item,
            index: count
          }
        })
      }
    }) : []

    return {
      ...prevState,
      searchResults,
      count,
      hasRequired,
      didSearch: ! nextProps.isSearching && prevState.searchText.length > 0
    }
  }

  componentDidMount(){
    document.addEventListener("keydown", this.handleKeyPress, false)
  }

  componentWillUnmount(){
    document.removeEventListener("keydown", this.handleKeyPress, false)
  }

  componentDidUpdate(prevProps) {
    const { mobileSearchVisible } = this.props

    if (prevProps.mobileSearchVisible !==  mobileSearchVisible && mobileSearchVisible ) {
      this.searchInput && this.searchInput.current.focus()
    }
  }

  handleBlur() {
    if (this.state.selected === 0) {
      this.setState({ isActive: false })
    }
  }

  handleKeyPress(e) {
    const { selected, searchResults, count, searchText, isActive } = this.state
    const goUp = () => {
      e.preventDefault()
      selected > 0 && this.setState({ selected: selected -1 })
      selected === 1 && this.searchInput && this.searchInput.current.focus()
    }

    const goDown = () => {
      e.preventDefault()
      count > selected && this.setState({ selected: selected + 1})
    }

    if (isActive) {
      switch(e.key) {
        case 'Enter':
          if (window.location.href.indexOf('/soeg') > -1) {
            this.setState({ isActive: false })
          } else {
            window.location.href = `https://opskrifter.secure:5002/soeg#?${encodeURI(searchText)}`
          }
          break

        case 'Escape':
          this.hideSearch()
          break

        case 'ArrowDown':
          searchResults.length && goDown()
          break

        case 'ArrowUp':
          searchResults.length && goUp()
          break

        case 'Tab':
          e.shiftKey ? goUp() : goDown()
          break
      }
    }
  }

  handleInputChange(e) {
    const value = e.target.value

    this.setState({ searchText: value, selected: 0, isActive: true })

    clearTimeout(this.timer)

    this.timer = setTimeout(() => {
      this.props.onSearchInput(value)
    }, 350)
  }

  hideSearch() {
    this.props.hideSearch()
    this.setState({ searchText: '', selected: 0, isActive: false })
  }

  render() {
    const { onSearchItemClick, isSearching, searchFailed } = this.props
    const { searchText, selected, searchResults, hasRequired, didSearch, isActive } = this.state
    const clearWrapClass = isActive ? 'coop-search__clear-wrap--has-input' : ''
    const maxHeightStyle = window.innerWidth < 768 ? { maxHeight: window.innerHeight - 60 } : {}

    return (
      <Fragment>
        <div className='coop-search'>
          <img
            src={getImagePath('search.svg')}
            alt="søg"
            className='coop-search__icon'
          />

          <input
            id='coop-search__input'
            className='coop-search__input'
            placeholder="Søg..."
            onClick={this.handleInputChange}
            onChange={this.handleInputChange}
            onBlur={this.handleBlur}
            value={searchText}
            ref={this.searchInput}
          />

          { isActive &&
            <Fragment>
              { isSearching &&
                <Loader
                  options={{
                    radius: 4,
                    length: 4,
                    width: 2,
                    lines: 10,
                    color: '#555',
                    left: 'auto',
                    className: 'coop-search__loader'
                  }}
                />
              }

              <div className='coop-search__results' style={maxHeightStyle}>
                { didSearch && ! hasRequired && ! searchFailed &&
                  <div className='coop-search__fallback'>Ingen resultater</div>
                }

                { searchFailed &&
                  <div className='coop-search__fallback'>Der er sket en fejl</div>
                }

                { searchResults.map((s, i) =>
                  <div key={i} className='coop-search__section'>
                    <h4 className='coop-search__section-title'>{s.title}</h4>

                    <div className='coop-search__section-results'>
                      { s.results.map((r, e) =>
                        <SearchItem
                          {...r}
                          key={e}
                          onSearchItemClick={onSearchItemClick}
                          isSelected={selected === r.index}
                          index={r.index}
                        />
                      )}
                    </div>
                  </div>
                )}
              </div>
            </Fragment>
          }

          <div className={`coop-search__clear-wrap ${clearWrapClass}`} onClick={this.hideSearch}>
            <img
              src={getImagePath('times-black.svg')}
              className='coop-search__clear-image'
              alt="Luk"
            />
          </div>
        </div>

        <Backdrop
          visible={isActive && searchResults.length > 0}
          onClose={this.hideSearch}
        />
      </Fragment>
    )
  }
}

export default Search