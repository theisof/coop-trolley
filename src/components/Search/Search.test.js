import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json';
import Search from './Search'

it('Search renders correctly', () => {
  const tree = shallow( <Search />)

  expect(toJSON(tree)).toMatchSnapshot()
})
