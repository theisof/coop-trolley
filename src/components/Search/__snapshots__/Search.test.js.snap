// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<Fragment>
  <div
    className="coop-search"
  >
    <img
      alt="søg"
      className="coop-search__icon"
      src="https://cdn.jsdelivr.net/gh/theisof/coop-trolley@2.0.0/images/search.svg"
    />
    <input
      className="coop-search__input"
      id="coop-search__input"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      placeholder="Søg..."
      value=""
    />
    <div
      className="coop-search__clear-wrap "
      onClick={[Function]}
    >
      <img
        alt="Luk"
        className="coop-search__clear-image"
        src="https://cdn.jsdelivr.net/gh/theisof/coop-trolley@2.0.0/images/times-black.svg"
      />
    </div>
  </div>
  <Backdrop
    onClose={[Function]}
    visible={false}
  />
</Fragment>
`;
